impl/src/main/java/org/jboss/weld/util/Beans.java

```diff
@@ -640,4 +644,44 @@ public static boolean hasBuiltinScope(Bean<?> bean) {
         manager.getServices().get(SlimAnnotatedTypeStore.class).put(implementationClass.slim());
         return implementationClass;
     }
+
+    static ImmutableSet.Builder<Type> omitIllegalBeanTypes(Set<Type> types, EnhancedAnnotated<?, ?> annotated) {
+        ImmutableSet.Builder<Type> builder = ImmutableSet.builder();
+        for (Type type : types) {
+            if (isIllegalBeanType(type)) {
+                MetadataLogger.LOG.illegalBeanTypeIgnored(type, annotated);
+            } else {
+                builder.add(type);
+            }
+        }
+        return builder;
+    }
+
+    /**
+     *
+     * @param beanType
+     * @return <code>true</code> if the given type is not a legal bean type, <code>false</code> otherwise
+     */
+    public static boolean isIllegalBeanType(Type beanType) {
```

Comment:
What about `Foo<Foo<?>>`? The spec does not forbid these types but it should as such type won't be assignable anywhere anyway. Therefore, we should probably process this recursively. That is what we do for events anyway.